[Yarr] Improve processing of \s* and related simple Character Class atoms

msaboff opened the pull request:
#### 33a069473f47b0242c67a1a4f9ae3a0ad249480d
<pre>
[Yarr] Improve processing of \s* and related simple Character Class atoms
<a href="https://bugs.webkit.org/show_bug.cgi?id=285121">https://bugs.webkit.org/show_bug.cgi?id=285121</a>
<a href="https://rdar.apple.com/141967884">rdar://141967884</a>

Reviewed by Yusuke Suzuki.

This has three fixes to improve the processing of simple character classes like \s or [abc].

1. Added code to YarrGenerator::matchCharacterClass() and related methods to generate a single
   branch instead of a branch over a branch.  e.g
     Before:
       &lt;112&gt; 0x11aabce30:    b.eq     0x11aabce38 -&gt; &lt;120&gt;
       &lt;116&gt; 0x11aabce34:    b        0x11aabce44 -&gt; &lt;132&gt;
       &lt;120&gt; 0x11aabce38:    ...
     After:
       &lt;112&gt; 0x11aabce30:    b.ne     0x11aabce38 -&gt; &lt;128&gt;
       &lt;126&gt; 0x11aabce38:    ...
   Part of this change can be used for future branch over branch elimination.
2. Fixed an issue in CharacterClass::copyOnly8BitCharacterData() where we didn&apos;t copy the table
   data.  This applies to the built-in whitespace and word classes.
3. Made a minor fix in the whitespace character class generation to turn the separate characters
   0x2028, 0x2029 into the range 0x2028..0x2029.

Added a micro benchmark to measure the perf improvement of \s+ for both 8 bit and 16 bit string.
The speedup for that micro benchmark is:
                              Baseline            NoBrAroundBr

regexp-greedy-whitespace  60.3322+-0.5647   ^   51.9861+-1.1278    ^ definitely 1.1605x faster

&lt;geometric&gt;               60.3322+-0.5647   ^   51.9861+-1.1278    ^ definitely 1.1605x faster

* JSTests/microbenchmarks/regexp-greedy-whitespace.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::CharacterClass::copyOnly8BitCharacterData):
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::CharacterClass::CharacterClass):
* Source/JavaScriptCore/yarr/create_regex_tables:

Canonical link: <a href="https://commits.webkit.org/288284@main">https://commits.webkit.org/288284@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/51d500c579630fc697b28b81a944f350059598c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82351 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2015 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36430 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87483 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33412 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2086 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9899 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64181 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21932 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85420 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1468 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74932 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44459 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1369 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29115 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32453 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75339 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72660 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29743 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88839 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81405 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9657 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6915 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72582 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9883 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70746 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71799 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17891 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15935 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14944 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1014 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9610 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15131 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103818 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9484 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25195 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12950 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11254 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -382,6 +382,76 @@ class YarrGenerator final : public YarrJITInfo {
+
+    private:
+        MacroAssembler::JumpList* m_matchSucceededTargets;
+        MacroAssembler::JumpList* m_matchFailedTargets;
```
Let's init them with `nullptr` by default via `{ nullptr }`

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.h:
```diff
@@ -85,6 +85,8 @@ inline CharacterClassWidths& operator|=(CharacterClassWidths& lhs, CharacterClas
 struct CharacterClass {
     WTF_MAKE_TZONE_ALLOCATED(CharacterClass);
 public:
+    typedef const char* Table;
```
Use `using`.

msaboff replied:
`using` used.

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -4659,6 +4748,7 @@ class YarrGenerator final : public YarrJITInfo {
         m_jit.emitFunctionEpilogue();
 #elif CPU(ARM64)
         if (mayCall()) {
+
```
Remove this line.

msaboff replied:
Done.

Constellation commented on Source/JavaScriptCore/yarr/YarrJIT.cpp:
```diff
@@ -4614,7 +4703,7 @@ class YarrGenerator final : public YarrJITInfo {
             m_jit.move(MacroAssembler::TrustedImm32(0xdc00), m_regs.trailingSurrogateTag);
         }
-#elif CPU(ARM_THUMB2)
+#elif CPU(ARM_THUMB2
```
It needs `)`

msaboff replied:
Fixed.

Constellation reviewed (approved):
r=me

msaboff committed:
[Yarr] Improve processing of \s* and related simple Character Class atoms
https://bugs.webkit.org/show_bug.cgi?id=285121
rdar://141967884

Reviewed by Yusuke Suzuki.

This has three fixes to improve the processing of simple character classes like \s or [abc].

1. Added code to YarrGenerator::matchCharacterClass() and related methods to generate a single
   branch instead of a branch over a branch.  e.g
     Before:
       <112> 0x11aabce30:    b.eq     0x11aabce38 -> <120>
       <116> 0x11aabce34:    b        0x11aabce44 -> <132>
       <120> 0x11aabce38:    ...
     After:
       <112> 0x11aabce30:    b.ne     0x11aabce38 -> <128>
       <126> 0x11aabce38:    ...
   Part of this change can be used for future branch over branch elimination.
2. Fixed an issue in CharacterClass::copyOnly8BitCharacterData() where we didn't copy the table
   data.  This applies to the built-in whitespace and word classes.
3. Made a minor fix in the whitespace character class generation to turn the separate characters
   0x2028, 0x2029 into the range 0x2028..0x2029.

Added a micro benchmark to measure the perf improvement of \s+ for both 8 bit and 16 bit string.
The speedup for that micro benchmark is:
                              Baseline            NoBrAroundBr

regexp-greedy-whitespace  60.3322+-0.5647   ^   51.9861+-1.1278    ^ definitely 1.1605x faster

<geometric>               60.3322+-0.5647   ^   51.9861+-1.1278    ^ definitely 1.1605x faster

* JSTests/microbenchmarks/regexp-greedy-whitespace.js: Added.
(arrayToString):
(objectToString):
(dumpValue):
(compareArray):
(compareGroups):
(testRegExp):
(testRegExpSyntaxError):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::CharacterClass::copyOnly8BitCharacterData):
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::CharacterClass::CharacterClass):
* Source/JavaScriptCore/yarr/create_regex_tables:

Canonical link: https://commits.webkit.org/288284@main
